File: app/src/org/koreader/launcher/AssetsUtils.java
package org.koreader.launcher;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

import android.content.Context;
import android.content.res.AssetManager;


/* Utils to extract compressed assets from the asset loader.
   Heavily inspired by https://stackoverflow.com/a/27375602

   This file needs a refactor to work with kotlin */

class AssetsUtils {
    private static final int BASE_BUFFER_SIZE = 1024;
    private static final String TAG = "AssetsUtils";

    /* do not extract assets if the same version is already installed */
    static boolean isSameVersion(Context context, String zipFile) {
        final String new_version = getPackageRevision(zipFile);
        try {
            String output = context.getFilesDir().getAbsolutePath();
            FileReader fileReader = new FileReader(output + "/git-rev");
            BufferedReader bufferedReader = new BufferedReader(fileReader);
            String installed_version = bufferedReader.readLine();
            bufferedReader.close();
            if (new_version.equals(installed_version)) {
                Logger.INSTANCE.i("Skip installation for revision " + new_version);
                return true;
            } else {
                Logger.INSTANCE.i("Found new package revision " + new_version);
                return false;
            }
        } catch (Exception e) {
            Logger.INSTANCE.i("Found new package revision " + new_version);
            return false;
        }
    }

    /* get the first zip file inside the assets module */
    static String getZipFile(Context context) {
        AssetManager assetManager = context.getAssets();
        try {
            String[] assets = assetManager.list("module");
            if (assets != null) {
                for (String asset : assets) {
                    if (asset.endsWith(".zip")) {
                        return asset;
                    }
                }
            }
            return null;
        } catch (IOException e) {
            Logger.INSTANCE.e(TAG, "error listing assets:\n" + e.toString());
            return null;
        }
    }

    /* copy raw assets under assets/module */
    static boolean copyUncompressedAssets(Context context) {
        AssetManager assetManager = context.getAssets();
        String assets_dir = context.getFilesDir().getAbsolutePath();
        // llapp_main.lua is the entry point for frontend code.
        boolean entry_point = false;
        try {
            String[] assets = assetManager.list("module");
            if (assets != null) {
                for (String asset : assets) {
                    File file = new File(assets_dir, asset);
                    InputStream input = assetManager.open("module/" + asset);
                    OutputStream output = new FileOutputStream(file);
                    Logger.INSTANCE.d(TAG, "copying " + asset +
                        " to " + file.getAbsolutePath());
                    copyFile(input, output);
                    input.close();
                    output.flush();
                    output.close();
                    if ("llapp_main.lua".equals(asset)) {
                        entry_point = true;
                    }
                }
            }
        } catch (IOException e) {
            Logger.INSTANCE.e(TAG, "error with raw assets:\n" + e.toString());
            entry_point = false;
        }
        return entry_point;
    }

    /* unzip from stream */
    static void unzip(InputStream stream, String output) {
        byte[] buffer = new byte[BASE_BUFFER_SIZE * 512];
        try {
            ZipInputStream inputStream = new ZipInputStream(stream);
            ZipEntry zipEntry = null;

            while ((zipEntry = inputStream.getNextEntry()) != null) {
                Logger.INSTANCE.d(TAG, "unzipping " + zipEntry.getName());

                if (zipEntry.isDirectory()) {
                    dirChecker(output, zipEntry.getName());
                } else {
                    File f = new File(output, zipEntry.getName());
                    if (!f.exists()) {
                        boolean success = f.createNewFile();
                        if (!success) {
                            Logger.INSTANCE.w(TAG,
                                "Failed to create file " + f.getName());
                            continue;
                        }
                        FileOutputStream outputStream = new FileOutputStream(f);
                        int count;
                        while ((count = inputStream.read(buffer)) != -1) {
                            outputStream.write(buffer, 0, count);
                        }
                        inputStream.closeEntry();
                        outputStream.close();
                    }
                }
            }
            inputStream.close();
        } catch (Exception e) {
            Logger.INSTANCE.e(TAG, "error unzipping assets:\n" + e.toString());
        }
    }

    /* copy files from stream */
    private static void copyFile(InputStream input, OutputStream output) throws IOException {
        byte[] buffer = new byte[BASE_BUFFER_SIZE];
        int read;
        while((read = input.read(buffer)) != -1) {
            output.write(buffer, 0, read);
        }
    }

    /* create new folders on demand */
    private static void dirChecker(String path, String file) {
        File f = new File(path, file);
        if (!f.isDirectory()) {
            boolean success = f.mkdirs();
            if (!success) {
                Logger.INSTANCE.w(TAG, "failed to create folder " + f.getName());
            }
        }
    }

    /* get package revision from zipFile with scheme: name-revision.zip */
    private static String getPackageRevision(String zipFile) {
        String zipName = zipFile.replace(".zip","");
        String[] parts = zipName.split("-");
        return zipName.replace(parts[0] + "-", "");
    }
}

File: app/src/org/koreader/launcher/AssetsUtils.kt
package org.koreader.launcher

import java.io.BufferedReader
import java.io.File
import java.io.FileOutputStream
import java.io.FileReader
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.Locale
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream

import android.content.Context

/* Utils to extract compressed assets from the asset loader */

internal object AssetsUtils {
    private const val TAG = "AssetsUtils"
    private const val BASE_BUFFER_SIZE = 1024

    /* copy raw assets from the assets module */
    fun copyRawAssets(context: Context): Boolean {
        val assetManager = context.assets
        val assetsDir = context.filesDir.absolutePath
        var entryPoint = false
        try {
            val assets = assetManager.list("module")
            if (assets != null) {
                for (asset in assets) {
                    val file = File(assetsDir, asset)
                    val input = assetManager.open("module/$asset")
                    val output = FileOutputStream(file)
                    copyFile(input, output)
                    input.close()
                    output.flush()
                    output.close()
                    // llapp_main.lua is the entry point for frontend code.
                    if ("llapp_main.lua" == asset) {
                        entryPoint = true
                    }
                }
            }
        } catch (e: IOException) {
            entryPoint = false
            Logger.e(TAG, "error copying raw assets: $e")
        }
        return entryPoint
    }

    /* get the first zip file inside the assets module */
    fun getZipFromAsset(context: Context): String? {
        val assetManager = context.assets
        try {
            val assets = assetManager.list("module")
            if (assets != null) {
                for (asset in assets) {
                    if (asset.endsWith(".zip")) {
                        return asset
                    }
                }
            }
            return null
        } catch (e: Exception) {
            Logger.e(TAG, "error finding a zip in assets store: $e")
            return null
        }
    }

    /* check if installed files have the same revision as the asset zipfile */
    fun isSameVersion(context: Context, zipFile: String): Boolean {
        val newVersion = getPackageRevision(zipFile)
        try {
            val output = context.filesDir.absolutePath
            val fileReader = FileReader("$output/git-rev")
            val bufferedReader = BufferedReader(fileReader)
            val installedVersion = bufferedReader.readLine()
            bufferedReader.close()
            return if (newVersion == installedVersion) {
                Logger.i("Skip installation for revision $newVersion")
                true
            } else {
                Logger.i("Found new package revision $newVersion")
                false
            }
        } catch (e: Exception) {
            Logger.w(TAG, "$e")
            Logger.i("Found new package revision $newVersion")
            return false
        }
    }

    /**
     * unzip contents of a zip file from stream into an absolute path.
     *
     * @param stream the InputStream where the zip file is stored
     * @param output full path to the folder where the zip is extracted
     * @param overwrite files on destination
     * @return boolean ok
     */
    fun unzip(stream: InputStream, output: String, overwrite: Boolean = true): Boolean {
        var error = false
        var newFiles = 0
        var updatedFiles = 0
        var skippedFiles = 0
        try {
            val inputStream = ZipInputStream(stream)
            var zipEntry: ZipEntry? = inputStream.nextEntry
            while (zipEntry != null) {
                if (zipEntry.isDirectory) {
                    dirChecker(output, zipEntry.name)
                } else {
                    val f = File(output, zipEntry.name)
                    if (f.exists()) {
                        if (overwrite) {
                            updatedFiles = ++updatedFiles
                        } else {
                            skippedFiles = ++skippedFiles
                        }
                    } else {
                        newFiles = ++newFiles
                        if (!f.createNewFile()) {
                            Logger.e(TAG, "Failed to create file " + f.name)
                            error = true
                            continue
                        }
                    }
                    if (f.exists() and overwrite) {
                        FileOutputStream(f).use { target ->
                            val buffer = ByteArray(16 * BASE_BUFFER_SIZE)
                            var len = inputStream.read(buffer)
                            while (len != -1) {
                                target.write(buffer, 0, len)
                                len = inputStream.read(buffer)
                            }
                            inputStream.closeEntry()
                        }
                    }
                }
                zipEntry = inputStream.nextEntry
            }
            inputStream.close()
            Logger.i(String.format(Locale.US,
                "Assets extracted without errors: %d updated, %d skipped, %d new",
                updatedFiles, skippedFiles, newFiles))
            return !error
        } catch (e: Exception) {
            Logger.e(TAG, "Error extracting assets: $e")
            return false
        }
    }

    /* get package revision from zipFile name. Zips must use the scheme: name-revision.zip */
    private fun getPackageRevision(zipFile: String): String {
        val zipName = zipFile.replace(".zip", "")
        val parts = zipName.split("-".toRegex()).dropLastWhile{ it.isEmpty() }.toTypedArray()
        return zipName.replace(parts[0] + "-", "")
    }

    /* copy files from stream */
    @Throws(IOException::class)
    private fun copyFile(input: InputStream, output: OutputStream) {
        try {
            input.use { source ->
                output.use { target ->
                    source.copyTo(target, BASE_BUFFER_SIZE)
                }
            }
        } catch (e: Exception) {
            Logger.e(TAG, "Error copying file: $e")
        }
    }

    /* create new folders on demand */
    private fun dirChecker(path: String, file: String) {
        val f = File(path, file)
        if (!f.isDirectory) {
            val success = f.mkdirs()
            if (!success) {
                Logger.w(TAG, "failed to create folder " + f.name)
            }
        }
    }
}
